### PR TITLE
INTEL_fragment_shader_ordering support

### PR DESCRIPTION
**Pull request by Erfan-Ahmadi**

### CODE REVIEW

File: spirv_glsl.cpp [hunk 1]
```diff
@@ -778,16 +772,28 @@ void CompilerGLSL::emit_header()
 		{
 			if (options.version < 310)
 				SPIRV_CROSS_THROW("At least ESSL 3.10 required for fragment shader interlock.");
-			require_extension_internal("GL_NV_fragment_shader_interlock");
+			statement("#extension GL_NV_fragment_shader_interlock : enable");
```
Comment (HansKristian-Work) [inline spirv_glsl.cpp:775]: Is there any reason this is not using the require_extension_internal mechanism? I'd expect this code to request GL_ARB_fragment_shader_interlock, and then the code that emits extensions will handle the ifdefs. This is similar to how other conditional extensions do it.

File: spirv_glsl.cpp [hunk 2]
```diff
@@ -778,16 +772,28 @@ void CompilerGLSL::emit_header()
...
+			statement("#define SPIRV_Cross_beginInvocationInterlock() beginInvocationInterlockNV()");
+			statement("#define SPIRV_Cross_endInvocationInterlock() endInvocationInterlockNV()");
 		}
 		else
 		{
 			if (options.version < 420)
-				require_extension_internal("GL_ARB_shader_image_load_store");
-			require_extension_internal("GL_ARB_fragment_shader_interlock");
+				statement("#extension GL_ARB_shader_image_load_store : enable");
```
Comment (HansKristian-Work) [inline spirv_glsl.cpp:782]: This is problematic if some other code used require_extension_internal (earlier in emit_header, this is exactly the case), we'd end up with double declaration, so this should also be require_extension_internal.

File: spirv_glsl.cpp [hunk 3]
```diff
@@ -778,16 +772,28 @@ void CompilerGLSL::emit_header()
...
+
+		    statement("#ifdef GL_ARB_fragment_shader_interlock");
```
Comment (HansKristian-Work) [inline spirv_glsl.cpp:784]: These ifdef ladders can be handled in the code that emits extensions below. I'd prefer consistency in this case even if it's not strictly necessary for correctness.

File: spirv_glsl.cpp [hunk 4]
```diff
@@ -1041,14 +1047,32 @@ void CompilerGLSL::emit_header()
 		if (execution.flags.get(ExecutionModePostDepthCoverage))
 			inputs.push_back("post_depth_coverage");
 
+		if (execution.flags.get(ExecutionModePixelInterlockOrderedEXT) ||
```
Comment (HansKristian-Work) [inline spirv_glsl.cpp:1050]: Should probably combine this check into a bool so it's not redundantly checked below.

File: spirv_glsl.cpp [hunk 5]
```diff
@@ -1041,14 +1047,32 @@ void CompilerGLSL::emit_header()
...
+			execution.flags.get(ExecutionModePixelInterlockUnorderedEXT) ||
+			execution.flags.get(ExecutionModeSampleInterlockOrderedEXT) ||
+			execution.flags.get(ExecutionModeSampleInterlockUnorderedEXT))
+		{
+			statement("#if defined(GL_ARB_fragment_shader_interlock)");
+		}
+
 		if (execution.flags.get(ExecutionModePixelInterlockOrderedEXT))
-			inputs.push_back("pixel_interlock_ordered");
+			statement("layout(pixel_interlock_ordered) in;");
```
Comment (HansKristian-Work) [inline spirv_glsl.cpp:1059]: I assume it's allowed to declare:
```
layout(pixel_interlock_ordered) in;
layout(early_fragment_tests) in;
```
where both execution modes are set?

**Review by HansKristian-Work — CHANGES_REQUESTED** [review]
The approach seems fine, but some nits.

File: spirv_glsl.cpp (hunk 5 above)
Comment (Erfan-Ahmadi) [reply spirv_glsl.cpp:1059]: Yes, I've checked generated code, both modes are set in this case.

File: spirv_glsl.cpp (hunk 2 above)
Comment (Erfan-Ahmadi) [reply spirv_glsl.cpp:782]: You're right, require extension may be called for GL_ARB_shader_image_load_store earlier.

File: spirv_glsl.cpp (hunk 1 above)
Comment (devshgraphicsprogramming) [reply spirv_glsl.cpp:775]: can you give an example?

File: spirv_glsl.cpp (hunk 2 above)
Comment (devshgraphicsprogramming) [reply spirv_glsl.cpp:782]: why are we even concerned with `GL_ARB_shader_image_load_store` ?

**Pull request closed without merging**